Cannot find xcrun when building with Nix flakes on Darwin/MacOS

On latest version of llama.cpp, building from flake (ie `nix build github:ggerganov/llama.cpp`) produces error with latest commit (https://github.com/ggerganov/llama.cpp/commit/d01b3c4c32357567f3531d4e6ceffc5d23e87583 at time of writing). 

```
error: builder for '/nix/store/9qrdgaz8s1wfz9fwci01j2vi97y45cfy-llama-cpp-metalkit-0.0.0.drv' failed with exit code 1;
       last 25 log lines:
       > -- Generating done (0.0s)
       > CMake Warning:
       >   Manually-specified variables were not used by the project:
       >
       >     CMAKE_EXPORT_NO_PACKAGE_REGISTRY
       >     CMAKE_POLICY_DEFAULT_CMP0025
       >
       >
       > -- Build files have been written to: /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build
       > cmake: enabled parallel building
       > cmake: enabled parallel installing
       > Running phase: buildPhase
       > build flags: -j8
       > [1/132] Compiling Metal kernels
       > FAILED: bin/default.metallib /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/default.metallib
       > cd /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build && xcrun -sdk macosx metal -O3 -c /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.metal -o /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air && xcrun -sdk macosx metallib /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air -o /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/default.metallib && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-common.h && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.metal
       > /bin/sh: xcrun: command not found
       > [2/132] Building C object CMakeFiles/ggml.dir/ggml-alloc.c.o
       > [3/132] Building C object CMakeFiles/ggml.dir/ggml-backend.c.o
       > [4/132] Building C object CMakeFiles/ggml.dir/ggml-metal.m.o
       > [5/132] Building C object CMakeFiles/ggml.dir/ggml-quants.c.o
       > [6/132] Building C object CMakeFiles/ggml.dir/ggml.c.o
       > [7/132] Building CXX object CMakeFiles/llama.dir/unicode.cpp.o
       > [8/132] Building CXX object CMakeFiles/llama.dir/llama.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/9qrdgaz8s1wfz9fwci01j2vi97y45cfy-llama-cpp-metalkit-0.0.0.drv'.
```


The key error is `/bin/sh: xcrun: command not found` -- Nix provides a clean `$PATH` to CMake which does not include `/usr/bin` and so Cmake is unable to locate or run `/usr/bin/xcrun`. This behavior is specific to the combination of MacOS and Nix. 

This can be fixed by either manually adding `/usr/bin` to `$PATH` prior to configuration & build steps, or by using absolute path (`/usr/bin/xcrun`) and setting `dontFixCmake = true;` in package.nix. 

System details:
Nix 2.21 (sandbox = off), MacOS 14.4 Sonoma

--- a/.devops/nix/package.nix
+++ b/.devops/nix/package.nix
@@ -156,6 +156,8 @@ effectiveStdenv.mkDerivation (
     postPatch = ''
       substituteInPlace ./ggml-metal.m \
         --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
+      substituteInPlace ./ggml-metal.m \
+        --replace '[bundle pathForResource:@"default" ofType:@"metallib"];' "@\"$out/bin/default.metallib\";"
 
       # TODO: Package up each Python script or service appropriately.
       # If we were to migrate to buildPythonPackage and prepare the `pyproject.toml`,

--- a/.devops/nix/package.nix
+++ b/.devops/nix/package.nix
@@ -36,7 +36,8 @@
   # It's necessary to consistently use backendStdenv when building with CUDA support,
   # otherwise we get libstdc++ errors downstream.
   effectiveStdenv ? if useCuda then cudaPackages.backendStdenv else stdenv,
-  enableStatic ? effectiveStdenv.hostPlatform.isStatic
+  enableStatic ? effectiveStdenv.hostPlatform.isStatic,
+  precompileMetalShaders ? false
 }@inputs:
 
 let
@@ -166,12 +167,12 @@ effectiveStdenv.mkDerivation (
     '';
 
     # With PR#6015 https://github.com/ggerganov/llama.cpp/pull/6015,
-    # `default.metallib` is compiled with Metal compiler from XCode
+    # `default.metallib` may be compiled with Metal compiler from XCode
     # and we need to escape sandbox on MacOS to access Metal compiler.
     # `xcrun` is used find the path of the Metal compiler, which is varible
     # and not on $PATH
     # see https://github.com/ggerganov/llama.cpp/pull/6118 for discussion
-    __noChroot = effectiveStdenv.isDarwin && useMetalKit;
+    __noChroot = effectiveStdenv.isDarwin && useMetalKit && precompileMetalShaders;
 
     nativeBuildInputs =
       [
@@ -189,7 +190,7 @@ effectiveStdenv.mkDerivation (
       ]
       ++ optionals (effectiveStdenv.hostPlatform.isGnu && enableStatic) [
         glibc.static
-      ] ++ optionals (effectiveStdenv.isDarwin && useMetalKit) [
+      ] ++ optionals (effectiveStdenv.isDarwin && useMetalKit && precompileMetalShaders) [
         xcrunHost
       ];
 
@@ -235,7 +236,10 @@ effectiveStdenv.mkDerivation (
         # Should likely use `rocmPackages.clr.gpuTargets`.
         "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
       ]
-      ++ optionals useMetalKit [ (lib.cmakeFeature "CMAKE_C_FLAGS" "-D__ARM_FEATURE_DOTPROD=1") ];
+      ++ optionals useMetalKit [
+        (lib.cmakeFeature "CMAKE_C_FLAGS" "-D__ARM_FEATURE_DOTPROD=1")
+        (cmakeBool "LLAMA_METAL_EMBED_LIBRARY" (!precompileMetalShaders))
+      ];
 
     # TODO(SomeoneSerge): It's better to add proper install targets at the CMake level,
     # if they haven't been added yet.

--- a/.devops/nix/package.nix
+++ b/.devops/nix/package.nix
@@ -4,6 +4,7 @@
   config,
   stdenv,
   mkShell,
+  runCommand,
   cmake,
   ninja,
   pkg-config,
@@ -87,6 +88,11 @@ let
     ]
   );
 
+  xcrunHost = runCommand "xcrunHost" {} ''
+    mkdir -p $out/bin
+    ln -s /usr/bin/xcrun $out/bin
+  '';
+
   # apple_sdk is supposed to choose sane defaults, no need to handle isAarch64
   # separately
   darwinBuildInputs =
@@ -157,6 +163,14 @@ effectiveStdenv.mkDerivation (
       substituteInPlace ./*.py --replace "/usr/bin/env python" "${llama-python}/bin/python"
     '';
 
+    # With PR#6015 https://github.com/ggerganov/llama.cpp/pull/6015,
+    # `default.metallib` is compiled with Metal compiler from XCode
+    # and we need to escape sandbox on MacOS to access Metal compiler.
+    # `xcrun` is used find the path of the Metal compiler, which is varible
+    # and not on $PATH
+    # see https://github.com/ggerganov/llama.cpp/pull/6118 for discussion
+    __noChroot = effectiveStdenv.isDarwin && useMetalKit;
+
     nativeBuildInputs =
       [
         cmake
@@ -173,6 +187,8 @@ effectiveStdenv.mkDerivation (
       ]
       ++ optionals (effectiveStdenv.hostPlatform.isGnu && enableStatic) [
         glibc.static
+      ] ++ optionals (effectiveStdenv.isDarwin && useMetalKit) [
+        xcrunHost
       ];
 
     buildInputs =

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1260,6 +1260,12 @@ if (LLAMA_METAL)
             GROUP_READ
             WORLD_READ
         DESTINATION ${CMAKE_INSTALL_BINDIR})
+    if (NOT LLAMA_METAL_EMBED_LIBRARY)
+        install(
+            FILES ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/default.metallib
+            DESTINATION ${CMAKE_INSTALL_BINDIR}
+        )
+    endif()
 endif()
 
 #

